### PR TITLE
feat(compiler): generate export maps on build

### PR DESCRIPTION
**Pull request by tanner-reits**
## What is the current behavior?

Stencil does not generate export maps for any entry points.

GitHub Issue Number: N/A

## What is the new behavior?

Stencil will generate export maps at build time. This include the default entry point, entry points for the lazy-loader (if applicable), and each custom element (if applicable).

This behavior is gated behind a new config flag.

## Documentation

https://github.com/ionic-team/stencil-site/pull/1443

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Testing

Unit tests were added for the new behavior.

This feature can be manually tested by installing a build of this branch into the component starter and enabling the `generateExportMaps` config flag.

### CODE REVIEW

File: src/compiler/types/validate-primary-package-output-target.ts [hunk 1]
```diff
@@ -38,25 +49,32 @@ export const PRIMARY_PACKAGE_TARGET_CONFIGS = {
   dist: {
     getModulePath: (rootDir: string, outputTargetDir: string) =>
       normalizePath(relative(rootDir, join(outputTargetDir, 'index.js'))),
-    getTypesPath: (rootDir: string, outputTargetConfig: d.OutputTargetDist) =>
+    getTypesPath: (rootDir: string, outputTargetConfig: any) =>
```
Comment (christian-bromann) [inline src/compiler/types/validate-primary-package-output-target.ts:52]: Any reason we have to type this as `any`?

File: src/compiler/build/write-export-maps.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+ * the lazy loader (if applicable).
+ *
+ * @param config The validated Stencil config
+ * @param buildCtx The build context containing the components to generate export maps for
+ */
+export const writeExportMaps = (config: d.ValidatedConfig, buildCtx: d.BuildCtx) => {
+  const eligiblePrimaryTargets = config.outputTargets.filter(isEligiblePrimaryPackageOutputTarget);
+  if (eligiblePrimaryTargets.length > 0) {
+    const primaryTarget =
+      eligiblePrimaryTargets.find((o) => o.isPrimaryPackageOutputTarget) ?? eligiblePrimaryTargets[0];
```
Comment (christian-bromann) [inline src/compiler/build/write-export-maps.ts:24]: Any chance `eligiblePrimaryTargets[0]` can be `undefined` and requires us to have a check here?

File: src/compiler/build/write-export-maps.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    const outputTargetConfig = PRIMARY_PACKAGE_TARGET_CONFIGS[primaryTarget.type];
+
+    if (outputTargetConfig.getModulePath) {
```
Comment (christian-bromann) [inline src/compiler/build/write-export-maps.ts:27]: Even though we have types for this but who knows how user may circumvent this
```suggestion
    if (typeof outputTargetConfig.getModulePath === 'function') {
```

File: src/compiler/build/write-export-maps.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      const importPath = outputTargetConfig.getModulePath(config.rootDir, primaryTarget.dir);
+
+      if (importPath) {
+        execSync(`npm pkg set "exports[.][import]"="${importPath}"`);
+      }
+    }
+
+    if (outputTargetConfig.getMainPath) {
```
Comment (christian-bromann) [inline src/compiler/build/write-export-maps.ts:35]:
```suggestion
    if (typeof outputTargetConfig.getMainPath === 'function') {
```

File: src/compiler/build/write-export-maps.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+      const requirePath = outputTargetConfig.getMainPath(config.rootDir, primaryTarget.dir);
+
+      if (requirePath) {
+        execSync(`npm pkg set "exports[.][require]"="${requirePath}"`);
+      }
+    }
+
+    if (outputTargetConfig.getTypesPath) {
```
Comment (christian-bromann) [inline src/compiler/build/write-export-maps.ts:43]:
```suggestion
    if (typeof outputTargetConfig.getTypesPath === 'function') {
```

**Review by christian-bromann — COMMENTED** [review]
Some minor comments but overall very cool stuff!

File: src/compiler/types/validate-primary-package-output-target.ts (hunk 1 above)
Comment (tanner-reits) [reply src/compiler/types/validate-primary-package-output-target.ts:52]: Yes, this was intentional. I'll add a comment

File: src/compiler/build/write-export-maps.ts (hunk 2 above)
Comment (tanner-reits) [reply src/compiler/build/write-export-maps.ts:24]: We check prior to this line that there are records in the array

File: src/compiler/build/write-export-maps.ts (hunk 3 above)
Comment (tanner-reits) [reply src/compiler/build/write-export-maps.ts:27]: This is actually something we setup for primary output target validation. It's an internal config, not publicly exposed.

Comment (christian-bromann) [reply src/compiler/build/write-export-maps.ts:27]: no need for it then.

**Review by christian-bromann — APPROVED** [review]
LGTM 👍

**Pull request merged**